<h1 style='text-align: center'> SEAL Data Scientist Technical Test </h1>
<h3 style='text-align: center; font-style: italic'> Oleh: Garudapaksi Yusuf </h3>

---

## Section 1 | Data ETL

[Extract, transform, load](https://en.wikipedia.org/wiki/Extract,_transform,_load) (ETL) adalah tiga fase pemrosesan data yang terdiri dari:
1. **Ekstraksi dari sumber data**: dalam *project* ini, secara langsung data diekstraksi dari *source* yang disediakan,
2. **Transformasi dalam format yang sesuai**: karena data tersedia dalam format `.csv` dan `xlsx` —dan untuk memudahkan proses analisa—, proses transformasi data akan menggunakan library Pandas untuk mengubah format data dalam bentuk DataFrame, dan
3. ***Load* ke data *container***: berupa *variable instance* dalam bahasa pemrograman Python.

*Catatan:*
* *Source data dan instruksi: [Google Drive](https://drive.google.com/drive/folders/16e052sYcXIIhrOVQJ86G_S12SGqrlp96?usp=sharing).*
* *Proses ETL dilakukan untuk memenuhi instruksi yang dikutip dari `Data Scientist Techncial Test.xlsx`, yang tersedia pada folder `docs` dari repository ini dengan sedikit modifikasi untuk memudahkan pemahaman:*

    > Buatlah *table* seperti berikut dari data dinas sosial dan data *master* desa:

In [1]:
# Import library yang dibutuhkan untuk mengakses contoh table
import pandas as pd

# Mengakses contoh table yang tersedia dalam format .xlsx
bentuk_table = pd.read_excel(
    io='../docs/Data Scientist Techncial Test.xlsx', # Letak dan nama file dari contoh table
    sheet_name='bentuk table', # Nama sheet dari contoh table
)

In [2]:
# Pengaturan untuk library Pandas 
pd.set_option(
    'display.max_columns', # Batasan untuk jumlah kolom yang ditampilkan
    None # Dihilangkan karena beberapa kolom disembunyikan dengan tampilan '...' 
)

# Menampilkan keseluruhan isi dari bentuk_table
bentuk_table

id  kode_provinsi_y nama_provinsi  kode_kabupaten_kota  \
0   1               35    JAWA TIMUR                 3526   
1   2               35    JAWA TIMUR                 3526   
2   3               35    JAWA TIMUR                 3526   
3   4               35    JAWA TIMUR                 3526   
4   5               35    JAWA TIMUR                 3526   
5   6               35    JAWA TIMUR                 3526   
6   7               35    JAWA TIMUR                 3526   
7   8               35    JAWA TIMUR                 3526   
8   9               35    JAWA TIMUR                 3526   
9  10               35    JAWA TIMUR                 3526   

             kabupaten  nama_kabupaten_kota  bps_kode_kecamatan    kecamatan  \
0  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526140        GEGER   
1  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526070        GALIS   
2  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526070        GALIS   
3  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526140        GEGER   
4  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526090       TRAGAH   
5  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526010        KAMAL   
6  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526020       LABANG   
7  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526080  TANAH MERAH   
8  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526120       BURNEH   
9  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526140        GEGER   

  bps_nama_kecamatan  bps_kode_desa_kelurahan       kelurahan  \
0              GEGER               3526140002          LERPAK   
1              GALIS               3526070004           GALIS   
2              GALIS               3526070009         LONGKEK   
3              GEGER               3526140006          CAMPOR   
4             TRAGAH               3526090013          TRAGAH   
5              KAMAL               3526010003       BANYUAJUH   
6             LABANG               3526020003  SUKOLILO BARAT   
7        TANAH MERAH               3526080022           POTER   
8             BURNEH               3526120003          BURNEH   
9              GEGER               3526140012  BANYONENG LAOK   

  bps_nama_desa_kelurahan kemendagri_kode_kecamatan kemendagri_nama_kecamatan  \
0                  LERPAK                  35.26.06                     GEGER   
1                   GALIS                  35.26.18                     GALIS   
2                 LONGKEK                  35.26.18                     GALIS   
3                  CAMPOR                  35.26.06                     GEGER   
4                  TRAGAH                  35.26.14                    TRAGAH   
5              BANYU AJUH                  35.26.04                     KAMAL   
6          SUKOLILO BARAT                  35.26.12                    LABANG   
7                   POTER                  35.26.13               TANAH MERAH   
8                  BURNEH                  35.26.03                    BURNEH   
9          BANYONENG LAOK                  35.26.06                     GEGER   

  kemendagri_kode_desa_kelurahan kemendagri_nama_desa_kelurahan  \
0                  35.26.06.2005                         LERPAK   
1                  35.26.18.2007                          GALIS   
2                  35.26.18.2006                        LONGKEK   
3                  35.26.06.2010                         CAMPOR   
4                  35.26.14.2006                         TRAGAH   
5                  35.26.04.2002                      BANYUAJUH   
6                  35.26.12.2006                 SUKOLILO BARAT   
7                  35.26.13.2014                          POTER   
8                  35.26.03.2001                         BURNEH   
9                  35.26.06.2002                 BANYONENG LAOK   

    periode_update               nik                    nama  \
0  TRIWULAN I 2024  3526060000000000    

*Catatan*:
* *Indeks tabel yang ditampilkan mulai dari `0-4` pada contoh tabel di atas adalah indeks yang ditampilkan dari Pandas (menggunakan sistem [0-indexing](https://pandas.pydata.org/docs/user_guide/indexing.html) yang dimulai dari angka 0), sedangkan*
* *Kolom `id` dengan nilai yang hampir serupa —mulai dari `1-5`— merupakan indeks yang disediakan dari `bentuk_table`.*

Berikutnya, `bentuk_table` di atas akan dibandingkan dengan data dinas sosial dan data *master* desa yang tersedia pada folder `data`:

In [3]:
# Mengakses data dinas sosial dalam format .csv
data_dinsos_csv = pd.read_csv(
    filepath_or_buffer='../data/Data Dinas Sosial.csv',
    sep=';', # Character yang digunakan sebagai delimiter/separator
)

# Dan dalam format .xlsx
data_dinsos_xlsx = pd.read_excel(
    io='../data/Data Dinas Sosial.xlsx',
)

# Mengakses data master desa
data_master_desa = pd.read_csv(
    filepath_or_buffer='../data/masterdesa.csv',
    sep=';'
)

In [4]:
# Menampilkan 5 baris pertama dari 'Data Dinas Sosial.csv'
data_dinsos_csv.head(
    n=5
)

id  kode_provinsi nama_provinsi   periode_update          nik  \
0   1             35    Jawa Timur  Triwulan I 2024  3,52606E+15   
1   2             35    Jawa Timur  Triwulan I 2024  3,52618E+15   
2   3             35    Jawa Timur  Triwulan I 2024  3,52619E+15   
3   4             35    Jawa Timur  Triwulan I 2024  3,52606E+15   
4   5             35    Jawa Timur  Triwulan I 2024  3,52614E+15   

           nama             alamat    provinsi       kabupaten kecamatan  \
0     ABD HALIM        DSN GLIMBUR  JAWA TIMUR  KAB. BANGKALAN     GEGER   
1   ABD. ROHMAN  DSN. REMBAH GALIS  JAWA TIMUR  KAB. BANGKALAN     GALIS   
2  ABDUS SYUKUR  DSN. BARAT SUMBER  JAWA TIMUR  KAB. BANGKALAN     GALIS   
3        ACHMAD         DSN POCONG  JAWA TIMUR  KAB. BANGKALAN     GEGER   
4  ACHMAD RIDOI     DSN. BANGGUMOK  JAWA TIMUR  KAB. BANGKALAN    TRAGAH   

  kelurahan  tahun kategori jumlah satuan  
0    LERPAK   2024        -      -      -  
1     GALIS   2024        -      -      -  
2   LONGKEK   2024        -      -      -  
3    CAMPOR   2024        -      -      -  
4    TRAGAH   2024        -      -      -

In [5]:
# Menampilkan 5 baris pertama dari 'Data Dinas Sosial.xlsx'
data_dinsos_xlsx.head(
    n=5
)

id  kode_provinsi nama_provinsi   periode_update                    nik  \
0   1             35    Jawa Timur  Triwulan I 2024  352606000000000000000   
1   2             35    Jawa Timur  Triwulan I 2024  352618000000000000000   
2   3             35    Jawa Timur  Triwulan I 2024  352619000000000032768   
3   4             35    Jawa Timur  Triwulan I 2024  352606000000000000000   
4   5             35    Jawa Timur  Triwulan I 2024  352614000000000000000   

           nama             alamat    provinsi       kabupaten kecamatan  \
0     ABD HALIM        DSN GLIMBUR  JAWA TIMUR  KAB. BANGKALAN     GEGER   
1   ABD. ROHMAN  DSN. REMBAH GALIS  JAWA TIMUR  KAB. BANGKALAN     GALIS   
2  ABDUS SYUKUR  DSN. BARAT SUMBER  JAWA TIMUR  KAB. BANGKALAN     GALIS   
3        ACHMAD         DSN POCONG  JAWA TIMUR  KAB. BANGKALAN     GEGER   
4  ACHMAD RIDOI     DSN. BANGGUMOK  JAWA TIMUR  KAB. BANGKALAN    TRAGAH   

  kelurahan  tahun kategori jumlah satuan  
0    LERPAK   2024        -      -      -  
1     GALIS   2024        -      -      -  
2   LONGKEK   2024        -      -      -  
3    CAMPOR   2024        -      -      -  
4    TRAGAH   2024        -      -      -

**Observasi:**

Sekilas tampak bahwa isi dari kedua format dari data dinas sosial `dinsos` pada tiap kolom hampir serupa.
* Kecuali pada kolom `nik` yang ditampilkan dalam format *scientific* (menggunakan notasi `E` untuk menampilkan bilangan numerik) pada `dinsos_csv`, dan menggunakan format numerik pada `dinsos_xlsx`.
* Analisis berikut menggunakan fungsi `equals()` untuk menguji kesamaan isi pada tiap kolom dari kedua format data tersebut.

In [6]:
# Uji kesamaan isi pada tiap kolom
pd.DataFrame(
    data={
        # Antara 'data_dinsos_csv'
        'data_dinsos_csv': data_dinsos_csv.columns,
        'equal': [data_dinsos_csv[column_csv].equals(
            other=data_dinsos_xlsx[column_xlsx]
        )for column_csv, column_xlsx in zip(data_dinsos_csv.columns, data_dinsos_xlsx.columns)],
        # Dengan 'data_dinsos_xlsx' 
        'data_dinsos_xlsx': data_dinsos_xlsx.columns
    }
)

data_dinsos_csv  equal data_dinsos_xlsx
0               id   True               id
1    kode_provinsi   True    kode_provinsi
2    nama_provinsi   True    nama_provinsi
3   periode_update   True   periode_update
4              nik  False              nik
5             nama   True             nama
6           alamat   True           alamat
7         provinsi   True         provinsi
8        kabupaten   True        kabupaten
9        kecamatan   True        kecamatan
10       kelurahan   True        kelurahan
11           tahun   True            tahun
12        kategori   True         kategori
13          jumlah   True           jumlah
14          satuan   True           satuan

**Observasi:**
* Hasil pengujian menunjukkan bahwa isi tiap kolom dari kedua format data bersifat identik (fungsi `equals()` menghasilkan boolean `True` untuk semua kolom), dengan pengecualian untuk kolom `nik`.
* Karena format `nik` pada `data_dinsos_csv` yang lebih mendekati format pada `bentuk_table`, tahapan berikutnya akan mengubah nilai `nik` dari sumber data ini ke dalam format numerik.

### Data Transformation

Sebelum proses transformasi tersebut dimulai, pertama-tama perlu dilihat terlebih dahulu nilai-nilai yang ada pada kolom tersebut:

In [7]:
# Menunjukkan nilai unik pada kolom 'nik' dalam `dinsos_csv`
data_dinsos_csv['nik'].unique()

array(['3,52606E+15', '3,52618E+15', '3,52619E+15', '3,52614E+15',
       '3,52604E+15', '3,52613E+15', '3,52603E+15', '3,52608E+15',
       '3,52607E+15', '3,5261E+15', '3,52601E+15', '3,52605E+15',
       '3,57805E+15', '3,52616E+15', '3,57804E+15', '3,52612E+15',
       '3,52617E+15', '3,52666E+15', '3,51324E+15', '3,52602E+15',
       '3,51019E+15', '3,51005E+15', '3,51016E+15', '3,51007E+15',
       '3,52211E+15', '3,51006E+15', '3,51021E+15', '3,51012E+15',
       '3,51003E+15', '3,51017E+15', '3,51009E+15', '3,51022E+15',
       '3,51014E+15', '3,51011E+15', '3,5102E+15', '3,51013E+15',
       '3,51001E+15', '3,51024E+15', '3,51015E+15', '3,51008E+15',
       '3,51002E+15', '3,51004E+15', '3,5101E+15', '3,51518E+15',
       '3,50502E+15', '3,50507E+15', '3,50505E+15', '3,50503E+15',
       '3,50509E+15', '3,50516E+15', '3,50508E+15', '3,50504E+15',
       '3,50518E+15', '3,50506E+15', '3,5051E+15', '3,50519E+15',
       '3,50306E+15', '3,50601E+15', '3,5222E+15', '3,52204E+15',


**Observasi:**

Nilai pada kolom tersebut memiliki `dtype` atau tipe data `object`.
* Contoh format nilai yang sering ditemui pada kolom ini adalah `3,52606E+15`.
* Ditemukan juga beberapa format lain seperti `` `3527035109120002 `` dan `'3503125502060001`.

Untuk menyesuaikan nilai-nilai tersebut dengan notasi *scientific* yang umumnya berlaku, maka karakter-karakter berikut perlu dimodifikasi:

In [8]:
# Mengganti karakter tertentu pada kolom 'nik' dalam 'data_dinsos_csv' 
data_dinsos_csv['nik'] = data_dinsos_csv['nik'].replace(
    to_replace={
        ',': '.',
        '`': '',
        "'": ''
    },
    regex=True # Mengimplementasikan regular expression pada parameter 'to_replace'
)

Setelah menggunakan format yang sesuai, nilai-nilai yang masih menggunakan tipe data *object* dapat dikonversi ke dalam tipe data *numeric*:

In [9]:
# Mengonversi tipe data pada kolom 'nik' dalam 'data_dinsos_csv'
data_dinsos_csv['nik'] = pd.to_numeric( # Ke tipe data numeric
    arg=data_dinsos_csv['nik']
)

print(f"Tipe data pada kolom 'nik' dalam 'data_dinsos_csv': {data_dinsos_csv['nik'].dtype}")

Tipe data pada kolom 'nik' dalam 'data_dinsos_csv': float64


Terakhir, nilai yang sudah menggunakan tipe data *numeric* namun dalam bentuk `float` diubah menjadi `int`:

In [10]:
# Mengubah tipe data pada kolom 'nik' dalam 'data_dinsos_csv'
data_dinsos_csv['nik'] = data_dinsos_csv['nik'].astype( 
    dtype=int # Menjadi tipe data 'int'
)

Berikut ditunjukkan hasil dari transformasi data pada kolom `nik` dalam `data_dinsos_csv`:

In [11]:
data_dinsos_csv['nik'].unique()

array([ 3526060000000000,  3526180000000000,  3526190000000000,
        3526140000000000,  3526040000000000,  3526130000000000,
        3526030000000000,  3526080000000000,  3526070000000000,
        3526100000000000,  3526010000000000,  3526050000000000,
        3578050000000000,  3526160000000000,  3578040000000000,
        3526120000000000,  3526170000000000,  3526660000000000,
        3513240000000000,  3526020000000000,  3510190000000000,
        3510050000000000,  3510160000000000,  3510070000000000,
        3522110000000000,  3510060000000000,  3510210000000000,
        3510120000000000,  3510030000000000,  3510170000000000,
        3510090000000000,  3510220000000000,  3510140000000000,
        3510110000000000,  3510200000000000,  3510130000000000,
        3510010000000000,  3510240000000000,  3510150000000000,
        3510080000000000,  3510020000000000,  3510040000000000,
        3510100000000000,  3515180000000000,  3505020000000000,
        3505070000000000,  3505050000000

Data tersebut dapat dibandingkan dengan data serupa yang tersedia pada `data_dinsos_xlsx` dan `bentuk_table` sebagai acuan format yang hendak digunakan:

In [12]:
pd.DataFrame(
    data=[
        len(bentuk_table['nik'].astype(str)[0]),
        len(data_dinsos_csv['nik'].astype(str)[0]),
        len(data_dinsos_xlsx['nik'].astype(str)[0])
    ],
    index=[
        'nik_table',
        'nik_csv',
        'nik_xlsx'
    ],
    columns=[
        'len()'
    ]
)

len()
nik_table     16
nik_csv       16
nik_xlsx      21

*Catatan:*

*Fungsi `len()` digunakan untuk menghitung panjang atau jumlah item dalam suatu object.*
* *Ketika object tersebut berwujud string, maka fungsi ini akan menghitung jumlah karakter dalam string tersebut.* 
* *Sehingga kolom `nik` perlu diubah tipe datanya terlebih dahulu menjadi `str`.*
* *Data pertama pada kolom ini diakses dengan integer-based indexing `[0]` untuk kemudian dihitung panjang datanya dengan `len()`.*

*Dengan kata lain, fungsi ini menghitung jumlah digit dari data kolom `nik` pada tiap sumber data*

**Observasi:**

Nilai `nik` pada `data_dinsos_csv` adalah yang paling sesuai:
* Karena panjang datanya serupa dengan yang telah dicontohkan pada `bentuk_table`, dan
* Secara *domain knowledge* yang umum diketahui, `nik` atau Nomor Induk Kependudukan terdiri dari 16 digit angka.

### ETL Finalization

Berdasarkan penemuan di atas, diputuskan untuk menggunakan `dinsos.csv` sebagai format yang representatif untuk data dinas sosial. Juga ditampilkan sumber data lainnya yakni data master desa.

In [13]:
# Membuat variable instance baru 
data_dinas_sosial = data_dinsos_csv # Dari format data dinas sosial yang akan digunakan

data_dinas_sosial.head(
    n=5
)

id  kode_provinsi nama_provinsi   periode_update               nik  \
0   1             35    Jawa Timur  Triwulan I 2024  3526060000000000   
1   2             35    Jawa Timur  Triwulan I 2024  3526180000000000   
2   3             35    Jawa Timur  Triwulan I 2024  3526190000000000   
3   4             35    Jawa Timur  Triwulan I 2024  3526060000000000   
4   5             35    Jawa Timur  Triwulan I 2024  3526140000000000   

           nama             alamat    provinsi       kabupaten kecamatan  \
0     ABD HALIM        DSN GLIMBUR  JAWA TIMUR  KAB. BANGKALAN     GEGER   
1   ABD. ROHMAN  DSN. REMBAH GALIS  JAWA TIMUR  KAB. BANGKALAN     GALIS   
2  ABDUS SYUKUR  DSN. BARAT SUMBER  JAWA TIMUR  KAB. BANGKALAN     GALIS   
3        ACHMAD         DSN POCONG  JAWA TIMUR  KAB. BANGKALAN     GEGER   
4  ACHMAD RIDOI     DSN. BANGGUMOK  JAWA TIMUR  KAB. BANGKALAN    TRAGAH   

  kelurahan  tahun kategori jumlah satuan  
0    LERPAK   2024        -      -      -  
1     GALIS   2024        -      -      -  
2   LONGKEK   2024        -      -      -  
3    CAMPOR   2024        -      -      -  
4    TRAGAH   2024        -      -      -

In [14]:
# Menampilkan 5 baris pertama dari data master desa
data_master_desa.head(
    n=5
)

kode_provinsi nama_provinsi  kode_kabupaten_kota nama_kabupaten_kota  \
0             35    JAWA TIMUR                 3501   KABUPATEN PACITAN   
1             35    JAWA TIMUR                 3501   KABUPATEN PACITAN   
2             35    JAWA TIMUR                 3501   KABUPATEN PACITAN   
3             35    JAWA TIMUR                 3501   KABUPATEN PACITAN   
4             35    JAWA TIMUR                 3501   KABUPATEN PACITAN   

   bps_kode_kecamatan bps_nama_kecamatan  bps_kode_desa_kelurahan  \
0             3501010           DONOROJO               3501010001   
1             3501010           DONOROJO               3501010002   
2             3501010           DONOROJO               3501010003   
3             3501010           DONOROJO               3501010004   
4             3501010           DONOROJO               3501010005   

  bps_desa_kelurahan kemendagri_kode_kecamatan kemendagri_nama_kecamatan  \
0             WIDORO                  35.01.01                  DONOROJO   
1            SAWAHAN                  35.01.01                  DONOROJO   
2              KALAK                  35.01.01                  DONOROJO   
3            SENDANG                  35.01.01                  DONOROJO   
4              KLEPU                  35.01.01                  DONOROJO   

  kemendagri_kode_desa_kelurahan kemendagri_desa_kelurahan  
0                  35.01.01.2001                    WIDORO  
1                  35.01.01.2002                   SAWAHAN  
2                  35.01.01.2003                     KALAK  
3                  35.01.01.2004                   SENDANG  
4                  35.01.01.2005                     KLEPU

Setelah proses ETL berhasil dijalankan, tahapan-tahapan berikutnya untuk membuat *table* dengan format yang sesuai dapat dilakukan.

---
## Section 2 | Table Creation
### Data Comparison
Berikut disertakan DataFrame yang berisikan daftar nama kolom pada `bentuk_table` dengan `data_dinas_sosial` dan `data_master_desa`:

In [15]:
pd.DataFrame(
    # Label/nama kolom pada tiap data
    data=[
        bentuk_table.columns, 
        data_dinas_sosial.columns,
        data_master_desa.columns
    ],
    # Indeks untuk tiap daftar nama kolom
    index=[
        'bentuk_table',
        'data_dinas_sosial',
        'data_master_desa'
    ]
).T # Transpose DataFrame untuk memudahkan analisis

bentuk_table data_dinas_sosial  \
0                               id                id   
1                  kode_provinsi_y     kode_provinsi   
2                    nama_provinsi     nama_provinsi   
3              kode_kabupaten_kota    periode_update   
4                        kabupaten               nik   
5              nama_kabupaten_kota              nama   
6               bps_kode_kecamatan            alamat   
7                        kecamatan          provinsi   
8               bps_nama_kecamatan         kabupaten   
9          bps_kode_desa_kelurahan         kecamatan   
10                       kelurahan         kelurahan   
11         bps_nama_desa_kelurahan             tahun   
12       kemendagri_kode_kecamatan          kategori   
13       kemendagri_nama_kecamatan            jumlah   
14  kemendagri_kode_desa_kelurahan            satuan   
15  kemendagri_nama_desa_kelurahan              None   
16                  periode_update              None   
17                             nik              None   
18                            nama              None   
19                          alamat              None   
20                        kategori              None   
21                      kategori.1              None   
22                          jumlah              None   
23                          satuan              None   
24                           tahun              None   

                  data_master_desa  
0                    kode_provinsi  
1                    nama_provinsi  
2              kode_kabupaten_kota  
3              nama_kabupaten_kota  
4               bps_kode_kecamatan  
5               bps_nama_kecamatan  
6          bps_kode_desa_kelurahan  
7               bps_desa_kelurahan  
8        kemendagri_kode_kecamatan  
9        kemendagri_nama_kecamatan  
10  kemendagri_kode_desa_kelurahan  
11       kemendagri_desa_kelurahan  
12                            None  
13                            None  
14                            None  
15                            None  
16                            None  
17                            None  
18                            None  
19                            None  
20                            None  
21                            None  
22                            None  
23                            None  
24                            None

**Observasi:**
* Beberapa kolom yang dibutuhkan untuk membuat *table* sesuai acuan dari `bentuk_table` hanya ditemukan persis satu di `data_dinas_sosial` maupun `data_master_desa`.
    * Seperti kolom `id`, `kode_kabupaten_kota`, dan `bps_kode_kecamatan`.
    * Sehingga terdapat kemungkinan untuk melakukan penggabungan atau *merge* dari dua sumber data ini.
* Namun proses penggabungan DataFrame membutuhkan *key*: data beserta kolomnya yang identik pada DataFrame yang hendak digabungkan.
    * Meskipun kedua kolom `kode_provinsi` pada masing-masing sumber data identik, tetapi `bentuk_table` menunjukkan bahwa `kode_provinsi` yang digunakan memiliki imbuhan `_y` —menandakan bahwa kolom ini berasal dari DataFrame kedua yang hendak digabungkan—, sehingga kolom ini tidak dijadikan *key* untuk membuat tabel baru.
    * *Merge umumnya menambahkan imbuhan di akhir nama kolom untuk dua kolom dengan nama yang sama dan akan digabungkan, sehingga imbuhan tersebut menandakan dari DataFrame mana asalnya.*
    * Dengan demikian, kolom yang akan ditentukan sebagai `key` adalah:
        1. Kolom `nama_provinsi`,
        2. Pasangan kolom `kabupaten` dan `nama_kabupaten_kota`,
        3. Antara `kecamatan` dengan `bps_nama_kecamatan` atau `kemendagri_nama_kecamatan`, dan 
        4. Antara `kelurahan` dengan `bps_desa_kelurahan` atau `kemendagri_nama_desa_kelurahan`.
    * Khusus untuk nomor 3 dan 4, analisis berikut akan menentukan pasangan kolom mana yang akan digunakan sebagai `key` berikutnya:

In [16]:
# Jumlah dari elemen dalam kolom 'kecamatan' dan 'kelurahan'
kecamatan_len = len(data_dinas_sosial['kecamatan'])
kelurahan_len = len(data_dinas_sosial['kelurahan'])

# Jumlah dari elemen tiap kolom yang berada pada kolom 'bps_' dan 'kemendagri'
kecamatan_bps_len = len(data_dinas_sosial[data_dinas_sosial['kecamatan'].isin(values=data_master_desa['bps_nama_kecamatan'])])
kelurahan_bps_len = len(data_dinas_sosial[data_dinas_sosial['kelurahan'].isin(values=data_master_desa['bps_desa_kelurahan'])])
kecamatan_kemendagri_len = len(data_dinas_sosial[data_dinas_sosial['kecamatan'].isin(values=data_master_desa['kemendagri_nama_kecamatan'])])
kelurahan_kemendagri_len = len(data_dinas_sosial[data_dinas_sosial['kelurahan'].isin(values=data_master_desa['kemendagri_desa_kelurahan'])])

pd.DataFrame(
    data=[
        (kecamatan_len, kelurahan_len),
        (kecamatan_bps_len, kelurahan_bps_len),
        (kecamatan_kemendagri_len, kelurahan_kemendagri_len)
    ],
    index=[
        'len()',
        'isin(bps_)',
        'isin(kemendagri_)'
    ],
    # Nama untuk tiap kolom pada DataFrame instance berikut
    columns=[
        'kecamatan',
        'kelurahan'
    ]
)

kecamatan  kelurahan
len()                  11583      11583
isin(bps_)             10853      10713
isin(kemendagri_)      11387      11545

**Observasi:**

Jumlah elemen dari tiap kolom `kecamatan` dan `kelurahan` yang persis sama dengan elemen pada kolom `bps_` lebih sedikit daripada kolom `kemendagri_`.
* Hal tersebut nantinya berdampak pada jumlah *missing value* yang lebih banyak jika menggabungkan masing-masing kolom dengan `bps_`.
* Dengan mempertimbangkan prinsip [kualitas data](https://en.wikipedia.org/wiki/Data_quality) yang membandingkan antara kondisi data saat ini dengan kondisi yang dikehendaki —salah satunya adalah *free of defect*—, maka kolom `kemendagri_` yang akan digunakan sebagai key pada tahapan berikutnya untuk mengurangi *defect* berupa *missing value* sebanyak mungkin.

Sebelum melakukan *merging*, perlu diperhatikan juga instruksi berikutnya:
> Untuk *header* kolom menggunakan bentuk kecil semua dan isi data menggunakan huruf besar semua.
* *Header* dari kolom yang akan digabungkan sudah dalam bentuk kecil atau *lowercase*, tetapi
* Beberapa kolom seperti `nama_provinsi` dan `periode_update` dari `data_dinas_sosial` masih berisikan nilai dalam format *lowercase*, sehingga tahapan berikut akan mengubah formatnya menjadi *uppercase*:

In [17]:
# Mengubah isi dari tiap kolom menjadi uppercase
data_dinas_sosial['nama_provinsi'] = data_dinas_sosial['nama_provinsi'].str.upper()
data_dinas_sosial['periode_update'] = data_dinas_sosial['periode_update'].str.upper()

Juga ditemukan nilai pada *key* `kabupaten` yang tidak persis sama dengan `nama_kabupaten_kota`. 
* Hal ini dikarenakan isi dari `kabupaten` menggunakan format `KAB.` lalu diikuti dengan nama dari kabupaten tersebut.
* Format ini pun perlu disesuaikan dengan mengikuti acuan dari `bentuk_table`: 

In [18]:
# Mengganti tiap data pada kolom 'kabupaten'
data_dinas_sosial['kabupaten'] = data_dinas_sosial['kabupaten'].str.replace(
    pat='KAB.', # Data dengan pola yang ditentukan
    repl='KABUPATEN' # Diganti dengan pola berikut
)

### Data Merging
Setelah perbandingan dan penyesuaian pada tahapan sebelumnya dinilai memadai, berikutnya proses penggabungan kedua data dapat dilakukan:

In [19]:
# Menggabungkan data dengan 'database-style join'
table = pd.merge(
    left=data_dinas_sosial, # Data pertama/left yang akan digabungkan
    right=data_master_desa, # Dengan data kedua/right
    how='left', # Menggunakan key dari left dan mempertahankan urutan dari key tersebut
    left_on=['nama_provinsi', 'kabupaten', 'kecamatan', 'kelurahan'], # Kolom dari left yang hendak digabungkan sebagai key
    right_on=['nama_provinsi', 'nama_kabupaten_kota', 'kemendagri_nama_kecamatan', 'kemendagri_desa_kelurahan'] # Dengan key dari right
)

table.head(
    n=5
)

id  kode_provinsi_x nama_provinsi   periode_update               nik  \
0   1               35    JAWA TIMUR  TRIWULAN I 2024  3526060000000000   
1   2               35    JAWA TIMUR  TRIWULAN I 2024  3526180000000000   
2   3               35    JAWA TIMUR  TRIWULAN I 2024  3526190000000000   
3   4               35    JAWA TIMUR  TRIWULAN I 2024  3526060000000000   
4   5               35    JAWA TIMUR  TRIWULAN I 2024  3526140000000000   

           nama             alamat    provinsi            kabupaten kecamatan  \
0     ABD HALIM        DSN GLIMBUR  JAWA TIMUR  KABUPATEN BANGKALAN     GEGER   
1   ABD. ROHMAN  DSN. REMBAH GALIS  JAWA TIMUR  KABUPATEN BANGKALAN     GALIS   
2  ABDUS SYUKUR  DSN. BARAT SUMBER  JAWA TIMUR  KABUPATEN BANGKALAN     GALIS   
3        ACHMAD         DSN POCONG  JAWA TIMUR  KABUPATEN BANGKALAN     GEGER   
4  ACHMAD RIDOI     DSN. BANGGUMOK  JAWA TIMUR  KABUPATEN BANGKALAN    TRAGAH   

  kelurahan  tahun kategori jumlah satuan  kode_provinsi_y  \
0    LERPAK   2024        -      -      -             35.0   
1     GALIS   2024        -      -      -             35.0   
2   LONGKEK   2024        -      -      -             35.0   
3    CAMPOR   2024        -      -      -             35.0   
4    TRAGAH   2024        -      -      -             35.0   

   kode_kabupaten_kota  nama_kabupaten_kota  bps_kode_kecamatan  \
0               3526.0  KABUPATEN BANGKALAN           3526140.0   
1               3526.0  KABUPATEN BANGKALAN           3526070.0   
2               3526.0  KABUPATEN BANGKALAN           3526070.0   
3               3526.0  KABUPATEN BANGKALAN           3526140.0   
4               3526.0  KABUPATEN BANGKALAN           3526090.0   

  bps_nama_kecamatan  bps_kode_desa_kelurahan bps_desa_kelurahan  \
0              GEGER             3.526140e+09             LERPAK   
1              GALIS             3.526070e+09              GALIS   
2              GALIS             3.526070e+09            LONGKEK   
3              GEGER             3.526140e+09             CAMPOR   
4             TRAGAH             3.526090e+09             TRAGAH   

  kemendagri_kode_kecamatan kemendagri_nama_kecamatan  \
0                  35.26.06                     GEGER   
1                  35.26.18                     GALIS   
2                  35.26.18                     GALIS   
3                  35.26.06                     GEGER   
4                  35.26.14                    TRAGAH   

  kemendagri_kode_desa_kelurahan kemendagri_desa_kelurahan  
0                  35.26.06.2005                    LERPAK  
1                  35.26.18.2007                     GALIS  
2                  35.26.18.2006                   LONGKEK  
3                  35.26.06.2010                    CAMPOR  
4                  35.26.14.2006                    TRAGAH

## Section 3 | Data Cleansing
### Missing Value
* Sebelum membuat modifikasi lebih jauh pada `table` yang dibentuk, inspeksi dapat dilakukan untuk memeriksa keberadaan anomali dalam data.
* Selaras dengan instruksi berikutnya yang diberikan:

    > Lakukan *check* pada data tersebut, apakah ada kolom (dengan) nilai *null* atau tidak.

* Karena inspeksi ini akan dilakukan berulang kali, maka dibuat sebuah fungsi untuk mengurangi pengulangan kode Python yang sama berulang kali:

In [20]:
def check_null():
    """Fungsi untuk mendeteksi dan menampilkan nilai null pada table

    Returns:
        DataFrame: DataFrame instance yang berisikan jumlah missing value pada tiap kolom table
    """
    df_null = pd.DataFrame(
        data=table\
            # Mendeteksi missing value (NA) pada tiap kolom dalam 'table'
            .isna()\
                # Kemudian menghitung jumlah missing value tersebut
                .sum(),
    # Mengganti nama kolom untuk memudahkan pemahaman
    ).rename(
        columns={0: 'NA (n)'}
    )
    return df_null

In [21]:
# Deteksi pertama nilai null pada table
# Untuk mencari keberadaan anomali dalam data
check_null()

NA (n)
id                                   0
kode_provinsi_x                      0
nama_provinsi                        0
periode_update                       0
nik                                  0
nama                                 0
alamat                               0
provinsi                             0
kabupaten                            0
kecamatan                            0
kelurahan                            0
tahun                                0
kategori                             0
jumlah                               0
satuan                               0
kode_provinsi_y                    291
kode_kabupaten_kota                291
nama_kabupaten_kota                291
bps_kode_kecamatan                 291
bps_nama_kecamatan                 291
bps_kode_desa_kelurahan            291
bps_desa_kelurahan                 291
kemendagri_kode_kecamatan          291
kemendagri_nama_kecamatan          291
kemendagri_kode_desa_kelurahan     291
kemendagri_desa_kelurahan          291

**Observasi:**
* Kolom dengan nilai *null* atau NA adalah kolom-kolom yang berasal dari `data_master_desa` (mulai dari `kode_provinsi_y` hingga `kemendagri_desa_kelurahan`).
* Tampilan berikut menunjukkan lokasi *missing value* tersebut dalam bentuk tabel:

In [22]:
# Menunjukkan baris data dalam table dengan missing value
table[table.isna().any(axis=1)]

id  kode_provinsi_x nama_provinsi     periode_update  \
136      137               35    JAWA TIMUR    TRIWULAN I 2024   
344      345               35    JAWA TIMUR    TRIWULAN I 2024   
380      381               35    JAWA TIMUR    TRIWULAN I 2024   
385      386               35    JAWA TIMUR    TRIWULAN I 2024   
746      747               35    JAWA TIMUR    TRIWULAN I 2024   
...      ...              ...           ...                ...   
11192  11192               35    JAWA TIMUR  TRIWULAN III 2024   
11193  11193               35    JAWA TIMUR  TRIWULAN III 2024   
11195  11195               35    JAWA TIMUR  TRIWULAN III 2024   
11219  11219               35    JAWA TIMUR  TRIWULAN III 2024   
11232  11232               35    JAWA TIMUR  TRIWULAN III 2024   

                    nik                     nama  \
136    3510130000000000    MOHAMAT JONI SETIAWAN   
344    3522090000000000   AWALUDIN MOHAMAD IRFAN   
380    3522090000000000  M QUBAYLAS SAHRONA AHDA   
385    3522100000000000                  MAROMAH   
746    3509220000000000         MA'RIFATUL AINUN   
...                 ...                      ...   
11192  3529180000000000                     LANI   
11193  3529170000000000                    LASRI   
11195  3529170000000000                 MISNATUN   
11219  3529080000000000              RADIMANSYAH   
11232  3529080000000000                    SU'AT   

                                 alamat    provinsi             kabupaten  \
136          DUSUN JATISARI RT 01 RW 01  JAWA TIMUR  KABUPATEN BANYUWANGI   
344        DUSUN KEDUNGDANG RT 12 RW 05  JAWA TIMUR  KABUPATEN BOJONEGORO   
380             DUSUN JANAR RT 07 RW 03  JAWA TIMUR  KABUPATEN BOJONEGORO   
385        DUSUN KAPAL ETAN RT 04 RW 04  JAWA TIMUR  KABUPATEN BOJONEGORO   
746    JLN. KH. MOH YASIN  RT.003 RW.18  JAWA TIMUR      KABUPATEN JEMBER   
...                                 ...         ...                   ...   
11192                 DUSUN TANAH LESEK  JAWA TIMUR     KABUPATEN SUMENEP   
11193                     DUSUN MURA'AS  JAWA TIMUR     KABUPATEN SUMENEP   
11195          RT 006 RW 003 DSN KRAJAN  JAWA TIMUR     KABUPATEN SUMENEP   
11219          RT 013 RW 004 SUSUN SAWO  JAWA TIMUR     KABUPATEN SUMENEP   
11232         RT 024 RW 007 SUSUN BURET  JAWA TIMUR     KABUPATEN SUMENEP   

          kecamatan       kelurahan  tahun kategori jumlah satuan  \
136    BLIMBINGSARI            BOMO   2024        -      -      -   
344       KEPOHBARU           KEPOH   2024        -      -      -   
380       KEPOHBARU        NGLUMBER   2024        -      -      -   
385       KEPOHBARU        BUMIREJO   2024        -      -      -   
746      SUMBERSARI     BANGSALSARI   2024        -      -      -   
...             ...             ...    ...      ...    ...    ...   
11192     BATUPUTIH     JURUAN LAOK   2024        -      -      -   
11193     BATUPUTIH           BADUR   2024        -      -      -   
11195     BATUPUTIH  LARANGAN BARMA   2024        -      -      -   
11219   GILIGINTING         LOMBANG   2024        -      -      -   
11232   GILIGINTING         LOMBANG   2024        -      -      -   

       kode_provinsi_y  kode_kabupaten_kota nama_kabupaten_kota  \
136                NaN                  NaN                 NaN   
344                NaN                  NaN                 NaN   
380                NaN                  NaN                 NaN   
385                NaN                  NaN                 NaN   
746                NaN                  NaN                 NaN   
...                ...                  ...                 ...   
11192              NaN                  NaN                 NaN   
11193              NaN                  NaN                 NaN   
11195              NaN                  NaN                 NaN   
11219              NaN                  NaN                 NaN   
11232              NaN                  NaN                 NaN   

       bps_kode_kecamatan bps_nam

**Observasi**:
* Jumlah baris data `291` yang ditunjukkan pada tabel di atas sudah sesuai dengan yang dideteksi dari `check_null()`.
* Instruksi berikutnya memberikan arahan sebagai berikut:

    > Jika ada nilai *null*, lakukan *cleansing* data pada *table* yang sudah dilakukan *join* sehingga meminimalisir nilai *null*.

    > Buatlah *coding cleansing* dengan bahasa Python. Asumsikan akan melakukan *cleansing* data pada data yang jumlahnya besar dan beragam, sehingga buatlah algoritma *coding* yang bisa digunakan untuk banyak *case* data.
    
    > Buatlah penjelasan *cleansing* data yang Anda lakukan dari *problem* dan solusi yang diberikan.

* Sesuai petunjuk tersebut, beberapa kolom tersebut dapat diisi dengan menggunakan kolom serupa yang berasal dari `data_dinas_sosial` sebagai acuan —karena kolom-kolom ini terisi sempurna tanpa *missing value*—, sebagaimana ditunjukkan dalam bagian-bagian berikut:
    * Bagian pertama mendefinisikan fungsi yang diperlukan untuk mengganti nilai *null*, dan
    * Bagian kedua menerapkan fungsi tersebut pada pasangan kolom dengan nilai NA dan kolom tanpa nilai NA.

In [23]:
def replace_null(value, replace):
    """Fungsi untuk mengganti nilai missing value

    Args:
        value (DataFrame): baris data dengan missing value
        replace (DataFrame): baris data tanpa missing value

    Returns:
        replace: jika value adalah missing value (None atau pd.isna())
        value: jika value selain missing value
    """    
    if value is None or pd.isna(value):
        return replace
    return value

def apply_replace_null(column_missing, column_filled):
    """Fungsi untuk menerapkan 'replace_null()' pada kolom table dengan missing value

    Args:
        column_missing (str): nama kolom yang berisikan missing value
        column_filled (str): nama kolom digunakan sebagai acuan untuk mengisi missing value tersebut
    """    
    table[column_missing] = table[[column_missing, column_filled]]\
        .apply(
            func=lambda row: replace_null(value=row[column_missing], replace=row[column_filled]),
            # Menggunakan 'axis' kedua atau kolom untuk menerapkan 'func' parameter
            # Sehingga 'func' diterapkan pada tiap baris dalam kolom tersebut
            axis=1
        )

In [24]:
# Menjalankan 'apply_replace_null()' pada kolom table dengan missing value
apply_replace_null(column_missing='kode_provinsi_y', column_filled='kode_provinsi_x')
apply_replace_null(column_missing='nama_kabupaten_kota', column_filled='kabupaten')
apply_replace_null(column_missing='bps_nama_kecamatan', column_filled='kecamatan')
apply_replace_null(column_missing='kemendagri_nama_kecamatan', column_filled='kecamatan')
apply_replace_null(column_missing='bps_desa_kelurahan', column_filled='kelurahan')
apply_replace_null(column_missing='kemendagri_desa_kelurahan', column_filled='kelurahan')

In [25]:
# Deteksi kedua nilai null pada table
# Setelah menjalankan 'apply_replace_null()' pada kolom tertentu dalam table
check_null()

NA (n)
id                                   0
kode_provinsi_x                      0
nama_provinsi                        0
periode_update                       0
nik                                  0
nama                                 0
alamat                               0
provinsi                             0
kabupaten                            0
kecamatan                            0
kelurahan                            0
tahun                                0
kategori                             0
jumlah                               0
satuan                               0
kode_provinsi_y                      0
kode_kabupaten_kota                291
nama_kabupaten_kota                  0
bps_kode_kecamatan                 291
bps_nama_kecamatan                   0
bps_kode_desa_kelurahan            291
bps_desa_kelurahan                   0
kemendagri_kode_kecamatan          291
kemendagri_nama_kecamatan            0
kemendagri_kode_desa_kelurahan     291
kemendagri_desa_kelurahan            0

**Observasi**:
* Fungsi `apply_replace_null()` berhasil dijalankan, dan *missing value* pada kolom yang ditetapkan telah terisi.
* Seperti halnya pada bagian deteksi awal sebelumnya, berikut ditampilkan contoh baris data dengan *missing value* untuk melihat hasil dari `apply_replace_null()`:

In [26]:
# Contoh baris data dengan missing value
# Pada baris dengan nomor indeks 136
# Kemudian mengubah baris data dalam bentuk Series ke dalam format DataFrame
table.iloc[136]\
    .to_frame()\
        .T

id kode_provinsi_x nama_provinsi   periode_update               nik  \
136  137              35    JAWA TIMUR  TRIWULAN I 2024  3510130000000000   

                      nama                      alamat    provinsi  \
136  MOHAMAT JONI SETIAWAN  DUSUN JATISARI RT 01 RW 01  JAWA TIMUR   

                kabupaten     kecamatan kelurahan tahun kategori jumlah  \
136  KABUPATEN BANYUWANGI  BLIMBINGSARI      BOMO  2024        -      -   

    satuan kode_provinsi_y kode_kabupaten_kota   nama_kabupaten_kota  \
136      -            35.0                 NaN  KABUPATEN BANYUWANGI   

    bps_kode_kecamatan bps_nama_kecamatan bps_kode_desa_kelurahan  \
136                NaN       BLIMBINGSARI                     NaN   

    bps_desa_kelurahan kemendagri_kode_kecamatan kemendagri_nama_kecamatan  \
136               BOMO                       NaN              BLIMBINGSARI   

    kemendagri_kode_desa_kelurahan kemendagri_desa_kelurahan  
136                            NaN                      BOMO

**Observasi**:
* Kolom tertentu pada tabel yang sebelumnya memiliki missing value telah terisi sesuai kolom acuannya.
* Untuk mengisi nilai NA lainnya, fungsi `update()` dapat digunakan dengan memanfaatkan data acuan `data_master_desa`: 

In [27]:
# Melakukan modifikasi isi pada table
table.update(
    other=data_master_desa, # Menggunakan 'non-NA values' dari 'data_master_desa'
    overwrite=False # Melakukan 'update()' hanya pada nilai NA pada table
)

In [28]:
# Deteksi ketiga nilai null pada table
check_null()

NA (n)
id                                   0
kode_provinsi_x                      0
nama_provinsi                        0
periode_update                       0
nik                                  0
nama                                 0
alamat                               0
provinsi                             0
kabupaten                            0
kecamatan                            0
kelurahan                            0
tahun                                0
kategori                             0
jumlah                               0
satuan                               0
kode_provinsi_y                      0
kode_kabupaten_kota                 78
nama_kabupaten_kota                  0
bps_kode_kecamatan                  78
bps_nama_kecamatan                   0
bps_kode_desa_kelurahan             78
bps_desa_kelurahan                   0
kemendagri_kode_kecamatan           78
kemendagri_nama_kecamatan            0
kemendagri_kode_desa_kelurahan      78
kemendagri_desa_kelurahan            0

**Observasi**:
* Fungsi `update()` juga berhasil dijalankan, dan meskipun proses ini tidak menghilangkan nilai *null* secara sepenuhnya, jumlah *missing value* kolom tertentu berkurang dari sebelumnya.
* Analisis berikut menunjukkan penyebab mengapa fungsi `update()` tidak dapat menggantikan nilai NA pada `table` secara sempurna:

In [29]:
table[table.isna().any(axis=1)]\
    .head(
        n=3 # Menampilkan 3 baris data dengan missing value untuk memudahkan pemahaman
    )

id  kode_provinsi_x nama_provinsi     periode_update  \
8705  8706               35    JAWA TIMUR  TRIWULAN III 2024   
8707  8708               35    JAWA TIMUR  TRIWULAN III 2024   
8708  8709               35    JAWA TIMUR  TRIWULAN III 2024   

                   nik                 nama  \
8705  3509190000000000        SIN MARCELINO   
8707  3509120000000000  SITI NURUL JUNAIDAH   
8708  3509130000000000        SLAMET ABIDIN   

                                                 alamat    provinsi  \
8705  JL. SULTAN AGUNG IX/19 LING. KAMPUNG LEDOK RT....  JAWA TIMUR   
8707                     DUSUN KRAJAN LOR RT.002 RW.002  JAWA TIMUR   
8708                    DUSUN GUMUK GEBANG RT 019 RW 22  JAWA TIMUR   

             kabupaten  kecamatan   kelurahan  tahun kategori jumlah satuan  \
8705  KABUPATEN JEMBER  KALIWATES      GELANG   2024        -      -      -   
8707  KABUPATEN JEMBER     AMBULU  TEGALBESAR   2024        -      -      -   
8708  KABUPATEN JEMBER  RAMBIPUJI   SUMBEREJO   2024        -      -      -   

      kode_provinsi_y  kode_kabupaten_kota nama_kabupaten_kota  \
8705             35.0                  NaN    KABUPATEN JEMBER   
8707             35.0                  NaN    KABUPATEN JEMBER   
8708             35.0                  NaN    KABUPATEN JEMBER   

      bps_kode_kecamatan bps_nama_kecamatan  bps_kode_desa_kelurahan  \
8705                 NaN          KALIWATES                      NaN   
8707                 NaN             AMBULU                      NaN   
8708                 NaN          RAMBIPUJI                      NaN   

     bps_desa_kelurahan kemendagri_kode_kecamatan kemendagri_nama_kecamatan  \
8705             GELANG                       NaN                 KALIWATES   
8707         TEGALBESAR                       NaN                    AMBULU   
8708          SUMBEREJO                       NaN                 RAMBIPUJI   

     kemendagri_kode_desa_kelurahan kemendagri_desa_kelurahan  
8705                            NaN                    GELANG  
8707                            NaN                TEGALBESAR  
8708                            NaN                 SUMBEREJO

In [30]:
# Membandingkan data pada table yang berisikan missing value
data_master_desa[ # Dengan data serupa pada `data_master_desa`
    # Menggunakan baris pertama pada table dengan missing value sebagai contoh
    # Dan menggunakan data dari kolom 'kemendagri_'
    (data_master_desa['nama_kabupaten_kota'] == 'KABUPATEN JEMBER') & 
    (data_master_desa['kemendagri_nama_kecamatan'] == 'KALIWATES') &
    (data_master_desa['kemendagri_desa_kelurahan'] == 'GELANG')
]

Empty DataFrame
Columns: [kode_provinsi, nama_provinsi, kode_kabupaten_kota, nama_kabupaten_kota, bps_kode_kecamatan, bps_nama_kecamatan, bps_kode_desa_kelurahan, bps_desa_kelurahan, kemendagri_kode_kecamatan, kemendagri_nama_kecamatan, kemendagri_kode_desa_kelurahan, kemendagri_desa_kelurahan]
Index: []

In [31]:
data_master_desa[
    # Serta menggunakan data dari kolom `bps_`
    (data_master_desa['nama_kabupaten_kota'] == 'KABUPATEN JEMBER') & 
    (data_master_desa['bps_nama_kecamatan'] == 'KALIWATES') &
    (data_master_desa['bps_desa_kelurahan'] == 'GELANG')
]

Empty DataFrame
Columns: [kode_provinsi, nama_provinsi, kode_kabupaten_kota, nama_kabupaten_kota, bps_kode_kecamatan, bps_nama_kecamatan, bps_kode_desa_kelurahan, bps_desa_kelurahan, kemendagri_kode_kecamatan, kemendagri_nama_kecamatan, kemendagri_kode_desa_kelurahan, kemendagri_desa_kelurahan]
Index: []

**Observasi:**
* Kedua [boolean indexing](https://pandas.pydata.org/docs/user_guide/10min.html#selection) di atas menunjukkan bahwa terdapat beberapa *key* dari `table` yang tidak ditemukan pada `data_master_desa`.
* Hal ini yang menyebabkan munculnya *missing value* ketika proses *merging* antara `data_dinas_sosial` dan `data_master_desa` diterapkan.
* Dengan demikian, proses imputasi nilai *null* terakhir pada `table` tidak dapat dilakukan, selain dengan:
    1. Melakukan pemutakhiran `data_master_desa` agar nilai-nilai di dalamnya sesuai dengan `data_dinas_sosial`, atau
    2. Melakukan imputasi secara *manual* pada `table` dengan data baru untuk mengisi nilai NA tersebut.

### Table Formatting

* Sesuai dengan hasil analisis pada bagian `Data Comparison` sebelumnya, `kode_provinsi` yang digunakan adalah `kode_provinsi_y` dan `kode_provinsi_x` tidak digunakan dalam tabel yang baru.
* Nama dari beberapa kolom seperti `bps_desa_kelurahan`, `kemendagri_desa_kelurahan`, dan `kategori` perlu disesuaikan dengan format nama kolom yang digunakan pada `contoh_tabel`.
* Demikian juga dengan urutan nama kolom pada `tabel` mengikuti urutan nama kolom `bentuk_table`

In [32]:
table = table\
    .drop( # Menghapus kolom tertentu dari table
        columns='kode_provinsi_x'
    ).rename( # Mengganti nama kolom sesuai format pada `bentuk_table`
        columns={
            'bps_desa_kelurahan': 'bps_nama_desa_kelurahan',
            'kemendagri_desa_kelurahan': 'kemendagri_nama_desa_kelurahan',
            'kategori': 'kategori.1'
        }
    ).reindex( # Mengatur urutan dari table 
        labels=bentuk_table.columns, # Menggunakan daftar kolom dari 'bentuk_table'
        axis=1 # 'Axis' kedua atau kolom yang diurutkan sesuai parameter 'labels' sebelumnya
    )

Hasil dari proses penggabungan ini berupa *variabel instance* `table` yang dapat dibandingkan dengan `bentuk_table`:

In [33]:
# Menunjukkan table yang baru dibentuk
print('Table baru:')
display(
    table.head(
        n=3
    )
)

# Dengan contoh table dalam satu tampilan
print('Contoh table:')
display(
    bentuk_table.head(
        n=3
    )
)

Table baru:


id  kode_provinsi_y nama_provinsi  kode_kabupaten_kota  \
0   1             35.0    JAWA TIMUR               3526.0   
1   2             35.0    JAWA TIMUR               3526.0   
2   3             35.0    JAWA TIMUR               3526.0   

             kabupaten  nama_kabupaten_kota  bps_kode_kecamatan kecamatan  \
0  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN           3526140.0     GEGER   
1  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN           3526070.0     GALIS   
2  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN           3526070.0     GALIS   

  bps_nama_kecamatan  bps_kode_desa_kelurahan kelurahan  \
0              GEGER             3.526140e+09    LERPAK   
1              GALIS             3.526070e+09     GALIS   
2              GALIS             3.526070e+09   LONGKEK   

  bps_nama_desa_kelurahan kemendagri_kode_kecamatan kemendagri_nama_kecamatan  \
0                  LERPAK                  35.26.06                     GEGER   
1                   GALIS                  35.26.18                     GALIS   
2                 LONGKEK                  35.26.18                     GALIS   

  kemendagri_kode_desa_kelurahan kemendagri_nama_desa_kelurahan  \
0                  35.26.06.2005                         LERPAK   
1                  35.26.18.2007                          GALIS   
2                  35.26.18.2006                        LONGKEK   

    periode_update               nik          nama             alamat  \
0  TRIWULAN I 2024  3526060000000000     ABD HALIM        DSN GLIMBUR   
1  TRIWULAN I 2024  3526180000000000   ABD. ROHMAN  DSN. REMBAH GALIS   
2  TRIWULAN I 2024  3526190000000000  ABDUS SYUKUR  DSN. BARAT SUMBER   

   kategori kategori.1 jumlah satuan  tahun  
0       NaN          -      -      -   2024  
1       NaN          -      -      -   2024  
2       NaN          -      -      -   2024

Contoh table:


id  kode_provinsi_y nama_provinsi  kode_kabupaten_kota  \
0   1               35    JAWA TIMUR                 3526   
1   2               35    JAWA TIMUR                 3526   
2   3               35    JAWA TIMUR                 3526   

             kabupaten  nama_kabupaten_kota  bps_kode_kecamatan kecamatan  \
0  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526140     GEGER   
1  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526070     GALIS   
2  KABUPATEN BANGKALAN  KABUPATEN BANGKALAN             3526070     GALIS   

  bps_nama_kecamatan  bps_kode_desa_kelurahan kelurahan  \
0              GEGER               3526140002    LERPAK   
1              GALIS               3526070004     GALIS   
2              GALIS               3526070009   LONGKEK   

  bps_nama_desa_kelurahan kemendagri_kode_kecamatan kemendagri_nama_kecamatan  \
0                  LERPAK                  35.26.06                     GEGER   
1                   GALIS                  35.26.18                     GALIS   
2                 LONGKEK                  35.26.18                     GALIS   

  kemendagri_kode_desa_kelurahan kemendagri_nama_desa_kelurahan  \
0                  35.26.06.2005                         LERPAK   
1                  35.26.18.2007                          GALIS   
2                  35.26.18.2006                        LONGKEK   

    periode_update               nik          nama             alamat  \
0  TRIWULAN I 2024  3526060000000000     ABD HALIM        DSN GLIMBUR   
1  TRIWULAN I 2024  3526180000000000   ABD. ROHMAN  DSN. REMBAH GALIS   
2  TRIWULAN I 2024  3526190000000000  ABDUS SYUKUR  DSN. BARAT SUMBER   

   kategori kategori.1 jumlah satuan  tahun  
0       NaN          -      -      -   2024  
1       NaN          -      -      -   2024  
2       NaN          -      -      -   2024

**Observasi:**
Secara keseluruhan —mulai dari urutan dan nama kolom, isi, serta bentuk dari data— dapat dinyatakan bahwa `table` yang dibentuk telah menyerupai unsur-unsur yang dicontohkan dalam `bentuk_table`, kecuali pada poin-poin berikut:
1. Terkecuali isi dari kolom `kode_provinsi_y`, `kode_kabupaten_kota`, `bps_kode_kecamatan`, dan `bps_kode_desa_kelurahan` pada `table` yang memiliki sedikit perbedaan pada nilai dan format yang digunakan.
    * Perbedaan ini berasal dari keberadaan *missing value* `NaN` pada kolom-kolom ini sebelum dilakukan *cleansing*. [Karena `NaN` adalah bilangan bulat `float`](https://pandas.pydata.org/docs/dev/user_guide/integer_na.html), maka Pandas melakukan *enforcing* tipe data pada kolom tersebut yang seharusnya bertipe bilangan bulat *integer* menjadi *floating point*.
    * Karena `Missing Value` pada bagian sebelumnya berhasil dijalankan, maka langkah berikutnya cukup dengan mengubah tipe data `table` menjadi `integer`
2. Sedangkan `Nan` pada kolom `kategori` dikarenakan pengubahan nama kolom ini dari `kategori.1`.
    * Dan sebenarnya `kategori.1` tidak ditemukan pada `data_dinas_sosial` maupun `data_master_desa`. Sehingga penyesuaian format `table` mengikuti `bentuk_table` menciptakan kolom tambahan yang tidak berisikan data apapun.
    * Untuk menyesuaikan dengan instruksi `meminimalisir nilai null` sebelumnya, serta menjalankan prinsip [validitas data](https://en.wikipedia.org/wiki/Data_validation) agar struktur data yang dibuat tersusun secara logis dan konsisten, maka diputuskan untuk tidak menyertakan kolom ini pada `table`.

In [34]:
table = table\
    .convert_dtypes( # Mengubah tipe data pada table
        convert_integer=True # Menjadi integer, jika dimungkinkan
    ).drop( # Menghapus kolom tertentu dari table
        columns='kategori'
    ).rename( # Mengubah nama kolom tertentu pada table
        columns={
            'kategori.1': 'kategori'
        }
    )

In [35]:
# Menampilkan table yang telah dibentuk dan dibersihkan
table

id  kode_provinsi_y nama_provinsi  kode_kabupaten_kota  \
0          1               35    JAWA TIMUR                 3526   
1          2               35    JAWA TIMUR                 3526   
2          3               35    JAWA TIMUR                 3526   
3          4               35    JAWA TIMUR                 3526   
4          5               35    JAWA TIMUR                 3526   
...      ...              ...           ...                  ...   
11578  11579               35    JAWA TIMUR                 3504   
11579  11580               35    JAWA TIMUR                 3504   
11580  11581               35    JAWA TIMUR                 3504   
11581  11582               35    JAWA TIMUR                 3504   
11582  11583               35    JAWA TIMUR                 3504   

                   kabupaten    nama_kabupaten_kota  bps_kode_kecamatan  \
0        KABUPATEN BANGKALAN    KABUPATEN BANGKALAN             3526140   
1        KABUPATEN BANGKALAN    KABUPATEN BANGKALAN             3526070   
2        KABUPATEN BANGKALAN    KABUPATEN BANGKALAN             3526070   
3        KABUPATEN BANGKALAN    KABUPATEN BANGKALAN             3526140   
4        KABUPATEN BANGKALAN    KABUPATEN BANGKALAN             3526090   
...                      ...                    ...                 ...   
11578  KABUPATEN TULUNGAGUNG  KABUPATEN TULUNGAGUNG             3504160   
11579  KABUPATEN TULUNGAGUNG  KABUPATEN TULUNGAGUNG             3504110   
11580  KABUPATEN TULUNGAGUNG  KABUPATEN TULUNGAGUNG             3504180   
11581  KABUPATEN TULUNGAGUNG  KABUPATEN TULUNGAGUNG             3504030   
11582  KABUPATEN TULUNGAGUNG  KABUPATEN TULUNGAGUNG             3504130   

        kecamatan bps_nama_kecamatan  bps_kode_desa_kelurahan    kelurahan  \
0           GEGER              GEGER               3526140002       LERPAK   
1           GALIS              GALIS               3526070004        GALIS   
2           GALIS              GALIS               3526070009      LONGKEK   
3           GEGER              GEGER               3526140006       CAMPOR   
4          TRAGAH             TRAGAH               3526090013       TRAGAH   
...           ...                ...                      ...          ...   
11578      KAUMAN             KAUMAN               3504160009     MOJOSARI   
11579   BOYOLANGU          BOYOLANGU               3504110002  KENDALBULUR   
11580   PAGERWOJO         PAGER WOJO               3504180005       PENJOR   
11581       PAKEL              PAKEL               3504030007         SODO   
11582  KEDUNGWARU         KEDUNGWARU               3504130010   MANGUNSARI   

      bps_nama_desa_kelurahan kemendagri_kode_kecamatan  \
0                      LERPAK                  35.26.06   
1                       GALIS                  35.26.18   
2                     LONGKEK                  35.26.18   
3                      CAMPOR                  35.26.06   
4                      TRAGAH                  35.26.14   
...                       ...                       ...   
11578                MOJOSARI                  35.04.05   
11579             KENDALBULUR                  35.04.02   
11580                  PENJOR                  35.04.06   
11581                    SODO                  35.04.18   
11582              MANGUNSARI                  35.04.03   

      kemendagri_nama_kecamatan kemendagri_kode_desa_kelurahan  \
0                         GEGER                  35.26.06.2005   
1                         GALIS                  35.26.18.2007   
2                         GALIS                  35.26.18.2006   
3                         GEGER                  35.26.06.2010   
4                        TRAGAH                  35.26.14.2006   
...                         ...                            ...   
11578                    KAUMAN                  35.04.05.2008   
11579                 BOYOLANGU                  35.04.02.2006   
11580                 PAGERWOJO                  35.

Dimensi dari *Table* tersebut dapat dibandingkan dengan sumber data dan contoh *table*:

In [36]:
def view_shape(data):
    """Fungsi untuk melihat bentuk/dimensi dari data

    Args:
        data (DataFrame): input dalam format DataFrame

    Returns:
        DataFrame: output yang berisikan jumlah baris dan kolom dari input 
    """    
    df_shape = pd.DataFrame(
        data=[
            data.shape[0],
            data.shape[1]
        ],
        index=[
            'Baris',
            'Kolom'
        ],
        columns=[
            'Jumlah'
        ]
    )
    return df_shape

In [37]:
# Menjalankan 'view_shape()' pada tiap data
print('Table baru:')
display(
    view_shape(
        data=table
    )
)

print('Data dinas sosial:')
display(
    view_shape(
        data=data_dinas_sosial
    )
)

print('Contoh table:')
display(
    view_shape(
        data=bentuk_table
    )
)

Table baru:


Jumlah
Baris   11583
Kolom      24

Data dinas sosial:


Jumlah
Baris   11583
Kolom      15

Contoh table:


Jumlah
Baris      10
Kolom      25

**Observasi:**
* Jumlah baris `11583` dari `table` yang telah dibentuk dan dibersihkan sudah sesuai dengan jumlah baris `11583` pada sumber data utama `data_dinas_sosial`.
* Jumlah kolom `24` dari `table` tersebut sudah sesuai dengan jumlah kolom `25` pada acuan `bentuk_table`, setelah menghilangkan 1 kolom `kategori.1`.

`Table` tersebut dapat disimpan ke dalam format `csv` pada folder `data` dari repository ini sebagai dokumentasi hasil *project*.

In [38]:
# Menyimpan 'table' dalam format .csv
table.to_csv(
    path_or_buf='../portfolio/table.csv',
    index=False
)

---